File: src/macos/LanMessenger/Core/Networking/Media/RemoteSessionRegistry.swift
import Foundation

// The ~10 s accept window, one-in-flight-session-per-peer, and single-shot
// session_id lookup.
//
// The clock is injected, so the entire window-expiry behaviour is testable in
// zero wall-clock time. Without that seam the only way to test a 10 s window is
// to wait 10 s, which means in practice it never gets tested.

/// An invite that has been accepted and is waiting for its media connection.
struct RemoteAcceptWindow: Equatable {
    let sessionID: String
    let peerPublicKeyB64: String
    let peerIP: String
    let role: RemoteSessionRole
    let keys: RemoteSessionKeys
    let openedAt: Date
    fileprivate(set) var attached: Bool = false

    static func == (lhs: RemoteAcceptWindow, rhs: RemoteAcceptWindow) -> Bool {
        lhs.sessionID == rhs.sessionID && lhs.attached == rhs.attached
    }
}

final class RemoteSessionRegistry {

    enum OpenResult: Equatable {
        case opened
        case busy(existingSessionID: String)
    }

    enum AdmitResult: Equatable {
        case admitted(RemoteAcceptWindow)
        /// No window with that id, or it was opened by a different peer.
        case unknown
        case expired
        /// The window has already been used. A session_id is single-shot: a
        /// second media_attach on the same id is either a duplicate or an
        /// attacker racing the legitimate viewer, and neither should win.
        case alreadyAttached
    }

    private let now: () -> Date
    private let acceptWindow: TimeInterval
    private let lock = NSLock()

    private var windows: [String: RemoteAcceptWindow] = [:]      // keyed by sessionID
    private var inFlightByPeer: [String: String] = [:]           // peerKey → sessionID
    private var liveSessions: [String: MediaSession] = [:]       // sessionID → session

    init(now: @escaping () -> Date = Date.init, acceptWindow: TimeInterval = 10.0) {
        self.now = now
        self.acceptWindow = acceptWindow
    }

    // MARK: - Accept windows

    /// Opens a window after an invite is accepted. One session in flight per
    /// peer: a peer that can open windows without bound can make a host hold
    /// derived key material for every one of them.
    @discardableResult
    func open(sessionID: String,
              peerPublicKeyB64: String,
              peerIP: String,
              role: RemoteSessionRole,
              keys: RemoteSessionKeys) -> OpenResult {
        lock.lock(); defer { lock.unlock() }
        pruneLocked()

        if let existing = inFlightByPeer[peerPublicKeyB64], existing != sessionID {
            return .busy(existingSessionID: existing)
        }
        windows[sessionID] = RemoteAcceptWindow(
            sessionID: sessionID, peerPublicKeyB64: peerPublicKeyB64, peerIP: peerIP,
            role: role, keys: keys, openedAt: now())
        inFlightByPeer[peerPublicKeyB64] = sessionID
        return .opened
    }

    /// Consumes a window for an inbound `media_attach`.
    ///
    /// `peerPublicKeyB64` is checked as well as the id: a session id travels in
    /// plaintext on the wire, so any LAN host that sees one could otherwise
    /// attach to a window it did not open.
    func admit(sessionID: String, peerPublicKeyB64: String) -> AdmitResult {
        lock.lock(); defer { lock.unlock() }
        pruneLocked()

        guard var window = windows[sessionID],
              window.peerPublicKeyB64 == peerPublicKeyB64 else { return .unknown }
        guard !window.attached else { return .alreadyAttached }
        guard now().timeIntervalSince(window.openedAt) <= acceptWindow else {
            windows[sessionID] = nil
            inFlightByPeer[window.peerPublicKeyB64] = nil
            return .expired
        }
        window.attached = true
        windows[sessionID] = window
        return .admitted(window)
    }

    func cancel(sessionID: String) {
        lock.lock(); defer { lock.unlock() }
        if let window = windows.removeValue(forKey: sessionID) {
            inFlightByPeer[window.peerPublicKeyB64] = nil
        }
    }

    func hasWindow(sessionID: String) -> Bool {
        lock.lock(); defer { lock.unlock() }
        return windows[sessionID] != nil
    }

    func inFlightSessionID(forPeer peerPublicKeyB64: String) -> String? {
        lock.lock(); defer { lock.unlock() }
        pruneLocked()
        return inFlightByPeer[peerPublicKeyB64]
    }

    /// Drops windows that were never used. Called on every registry operation
    /// rather than from a timer — an expired window costs nothing to hold for a
    /// few extra seconds, and a timer here would be one more thing to starve.
    private func pruneLocked() {
        let cutoff = now().addingTimeInterval(-acceptWindow)
        for (id, window) in windows where !window.attached && window.openedAt < cutoff {
            windows[id] = nil
            inFlightByPeer[window.peerPublicKeyB64] = nil
        }
    }

    // MARK: - Live sessions

    func register(_ session: MediaSession) {
        lock.lock(); defer { lock.unlock() }
        liveSessions[session.sessionID] = session
    }

    func remove(sessionID: String) {
        lock.lock()
        let window = windows.removeValue(forKey: sessionID)
        if let key = window?.peerPublicKeyB64 { inFlightByPeer[key] = nil }
        liveSessions[sessionID] = nil
        lock.unlock()
    }

    func session(id: String) -> MediaSession? {
        lock.lock(); defer { lock.unlock() }
        return liveSessions[id]
    }

    var liveSessionCount: Int {
        lock.lock(); defer { lock.unlock() }
        return liveSessions.count
    }

    /// Closes everything. Reached from `NetworkCoordinator.stop()` so a detached
    /// socket can never outlive the network stack that produced it.
    func closeAll() {
        lock.lock()
        let sessions = Array(liveSessions.values)
        liveSessions.removeAll()
        windows.removeAll()
        inFlightByPeer.removeAll()
        lock.unlock()
        for session in sessions { session.stop() }
    }
}

File: src/windows-native/LanMessenger/Core/Networking/Media/RemoteSessionRegistry.cs
using LanMessenger.Core.Crypto;

namespace LanMessenger.Core.Networking.Media;

// The ~10 s accept window, one-in-flight-session-per-peer, and single-shot
// session_id lookup. Mirror of the macOS RemoteSessionRegistry.swift.
//
// The clock is injected, so the entire window-expiry behaviour is testable in
// zero wall-clock time. Without that seam the only way to test a 10 s window is
// to wait 10 s, which means in practice it never gets tested.

/// <summary>An invite that has been accepted and is waiting for its media connection.</summary>
public sealed class RemoteAcceptWindow(
    string sessionId, string peerPublicKeyB64, string peerIP,
    RemoteSessionRole role, RemoteSessionKeys keys, DateTime openedAt)
{
    public string SessionId { get; } = sessionId;
    public string PeerPublicKeyB64 { get; } = peerPublicKeyB64;
    public string PeerIP { get; } = peerIP;
    public RemoteSessionRole Role { get; } = role;
    public RemoteSessionKeys Keys { get; } = keys;
    public DateTime OpenedAt { get; } = openedAt;
    public bool Attached { get; internal set; }
}

public enum RemoteOpenKind { Opened, Busy }
public enum RemoteAdmitKind { Admitted, Unknown, Expired, AlreadyAttached }

public readonly record struct RemoteOpenResult(RemoteOpenKind Kind, string ExistingSessionId = "");
public readonly record struct RemoteAdmitResult(RemoteAdmitKind Kind, RemoteAcceptWindow? Window = null);

public sealed class RemoteSessionRegistry(Func<DateTime>? now = null, TimeSpan? acceptWindow = null)
{
    private readonly Func<DateTime> _now = now ?? (() => DateTime.UtcNow);
    private readonly TimeSpan _acceptWindow = acceptWindow ?? TimeSpan.FromSeconds(10);
    private readonly object _gate = new();

    private readonly Dictionary<string, RemoteAcceptWindow> _windows = [];   // by sessionId
    private readonly Dictionary<string, string> _inFlightByPeer = [];        // peerKey -> sessionId
    private readonly Dictionary<string, MediaSession> _liveSessions = [];    // sessionId -> session

    // ---- Accept windows ----------------------------------------------------

    /// <summary>Opens a window after an invite is accepted.</summary>
    /// <remarks>
    /// One session in flight per peer: a peer that can open windows without bound
    /// can make a host hold derived key material for every one of them.
    /// </remarks>
    public RemoteOpenResult Open(
        string sessionId, string peerPublicKeyB64, string peerIP,
        RemoteSessionRole role, RemoteSessionKeys keys)
    {
        lock (_gate)
        {
            PruneLocked();
            if (_inFlightByPeer.TryGetValue(peerPublicKeyB64, out var existing) && existing != sessionId)
                return new RemoteOpenResult(RemoteOpenKind.Busy, existing);

            _windows[sessionId] = new RemoteAcceptWindow(
                sessionId, peerPublicKeyB64, peerIP, role, keys, _now());
            _inFlightByPeer[peerPublicKeyB64] = sessionId;
            return new RemoteOpenResult(RemoteOpenKind.Opened);
        }
    }

    /// <summary>Consumes a window for an inbound media_attach.</summary>
    /// <remarks>
    /// The peer key is checked as well as the id: a session id travels in
    /// plaintext on the wire, so any LAN host that sees one could otherwise
    /// attach to a window it did not open.
    /// </remarks>
    public RemoteAdmitResult Admit(string sessionId, string peerPublicKeyB64)
    {
        lock (_gate)
        {
            PruneLocked();
            if (!_windows.TryGetValue(sessionId, out var window)
                || window.PeerPublicKeyB64 != peerPublicKeyB64)
                return new RemoteAdmitResult(RemoteAdmitKind.Unknown);
            if (window.Attached)
                return new RemoteAdmitResult(RemoteAdmitKind.AlreadyAttached);
            if (_now() - window.OpenedAt > _acceptWindow)
            {
                _windows.Remove(sessionId);
                _inFlightByPeer.Remove(window.PeerPublicKeyB64);
                return new RemoteAdmitResult(RemoteAdmitKind.Expired);
            }
            window.Attached = true;
            return new RemoteAdmitResult(RemoteAdmitKind.Admitted, window);
        }
    }

    public void Cancel(string sessionId)
    {
        lock (_gate)
        {
            if (_windows.Remove(sessionId, out var window))
                _inFlightByPeer.Remove(window.PeerPublicKeyB64);
        }
    }

    public bool HasWindow(string sessionId) { lock (_gate) return _windows.ContainsKey(sessionId); }

    public string? InFlightSessionId(string peerPublicKeyB64)
    {
        lock (_gate)
        {
            PruneLocked();
            return _inFlightByPeer.TryGetValue(peerPublicKeyB64, out var id) ? id : null;
        }
    }

    /// <summary>
    /// Drops windows that were never used. Called on every registry operation
    /// rather than from a timer — an expired window costs nothing to hold for a
    /// few extra seconds, and a timer here would be one more thing to starve.
    /// </summary>
    private void PruneLocked()
    {
        DateTime cutoff = _now() - _acceptWindow;
        foreach (var id in _windows.Where(kv => !kv.Value.Attached && kv.Value.OpenedAt < cutoff)
                                   .Select(kv => kv.Key).ToList())
        {
            if (_windows.Remove(id, out var window))
                _inFlightByPeer.Remove(window.PeerPublicKeyB64);
        }
    }

    // ---- Live sessions -----------------------------------------------------

    public void Register(MediaSession session) { lock (_gate) _liveSessions[session.SessionId] = session; }

    public void Remove(string sessionId)
    {
        lock (_gate)
        {
            if (_windows.Remove(sessionId, out var window))
                _inFlightByPeer.Remove(window.PeerPublicKeyB64);
            _liveSessions.Remove(sessionId);
        }
    }

    public MediaSession? Session(string sessionId)
    {
        lock (_gate) return _liveSessions.TryGetValue(sessionId, out var s) ? s : null;
    }

    public int LiveSessionCount { get { lock (_gate) return _liveSessions.Count; } }

    /// <summary>
    /// Closes everything. Reached from NetworkCoordinator.Stop() so a detached
    /// socket can never outlive the network stack that produced it.
    /// </summary>
    public void CloseAll()
    {
        List<MediaSession> sessions;
        lock (_gate)
        {
            sessions = [.. _liveSessions.Values];
            _liveSessions.Clear();
            _windows.Clear();
            _inFlightByPeer.Clear();
        }
        foreach (var session in sessions) session.Stop();
    }
}
